telluric/util/raster_utils.py

```diff
@@ -83,3 +88,243 @@ def convert_to_cog(source_file, destination_file, resampling=Resampling.gauss):
+                src.crs, dst_crs, *src.bounds)
+            dst_transform = Affine(
+                (xmax - xmin) / float(dst_width),
+                0, xmin, 0,
+                (ymin - ymax) / float(dst_height),
+                ymax
+            )
+
+        elif src_bounds or dst_bounds:
+            if not resolution:
+                raise ValueError('resolution is required when using src_bounds or dst_bounds')
+
+            if src_bounds:
+                xmin, ymin, xmax, ymax = transform_bounds(
+                    src.crs, dst_crs, *src_bounds)
+            else:
+                xmin, ymin, xmax, ymax = dst_bounds
+
+            dst_transform = Affine(resolution[0], 0, xmin, 0, -resolution[1], ymax)
+            dst_width = max(int(ceil((xmax - xmin) / resolution[0])), 1)
+            dst_height = max(int(ceil((ymax - ymin) / resolution[1])), 1)
+
+        else:
+            if src.transform.is_identity and src.gcps:
+                src_crs = src.gcps[1]
+                kwargs = {'gcps': src.gcps[0]}
+            else:
+                src_crs = src.crs
+                kwargs = src.bounds._asdict()
+            dst_transform, dst_width, dst_height = calcdt(
+                src_crs, dst_crs, src.width, src.height,
+                resolution=resolution, **kwargs)
+
+    elif dimensions:
+        # Same projection, different dimensions, calculate resolution.
+        dst_crs = src.crs
+        dst_width, dst_height = dimensions
+        dst_transform = Affine(
+            (r - l) / float(dst_width),
+            0, l, 0,
+            (b - t) / float(dst_height),
+            t
+        )
+
+    elif src_bounds or dst_bounds:
+        # Same projection, different dimensions and possibly
+        # different resolution.
+        if not resolution:
+            resolution = (src.transform.a, -src.transform.e)
+
+        dst_crs = src.crs
+        xmin, ymin, xmax, ymax = (src_bounds or dst_bounds)
+        dst_transform = Affine(resolution[0], 0, xmin, 0, -resolution[1], ymax)
+        dst_width = max(int(ceil((xmax - xmin) / resolution[0])), 1)
+        dst_height = max(int(ceil((ymax - ymin) / resolution[1])), 1)
+
+    elif resolution:
+        # Same projection, different resolution.
+        dst_crs = src.crs
+        dst_transform = Affine(resolution[0], 0, l, 0, -resolution[1], t)
+        dst_width = max(int(ceil((r - l) / resolution[0])), 1)
+        dst_height = max(int(ceil((t - b) / resolution[1])), 1)
+
+    else:
+        dst_crs = src.crs
+        dst_transform = src.transform
+        dst_width = src.width
+        dst_height = src.height
+
+    if target_aligned_pixels:
+        dst_transform, dst_width, dst_height = aligned_target(
+            dst_transform, dst_width, dst_height, resolution)
+
+    return dst_crs, dst_transform, dst_width, dst_height
+
+
+# Code was adapted from rasterio.rio.warp module
+def warp(source_file, destination_file, dst_crs=None, resolution=None, dimensions=None,
```

Comment:
What were the necessary changes here? Any reason not to use the upstream function?

Reply:
For a rationale of splitting original *rasterio.rio.warp* into two functions see my comment above. Also I'm not sure that this is good idea of using console utility (decorated by multiple `@click`).